src/core/telemetry/baseTelemetrySender.ts

```diff
@@ -0,0 +1,107 @@
+import { SenderData } from "./baseTelemetryReporter";
+
+export interface BaseTelemetryClient {
+  logEvent(eventName: string, data?: SenderData): void;
+  logException(exceptionName: Error, data?: SenderData): void;
+  flush(): void | Promise<void>;
+}
+
+enum InstantiationStatus {
+  NOT_INSTANTIATED,
+  INSTANTIATING,
+  INSTANTIATED,
+}
+
+export interface TelemetrySender {
+  instantiateSender(): void;
+}
+
+export class BaseTelemetrySender implements TelemetrySender {
+  // Whether or not the client has been instantiated
+  private _instantiationStatus: InstantiationStatus = InstantiationStatus.NOT_INSTANTIATED;
+  private _telemetryClient: BaseTelemetryClient | undefined;
+
+  // Queues used to store events until the sender is ready
+  private _eventQueue: Array<{ eventName: string; data: SenderData | undefined }> = [];
+  private _exceptionQueue: Array<{ exception: Error; data: SenderData | undefined }> = [];
+
+  // Necessary information to create a telemetry client
+  private _clientFactory: (key: string) => Promise<BaseTelemetryClient>;
+  private _key: string;
+
+  constructor(key: string, clientFactory: (key: string) => Promise<BaseTelemetryClient>) {
+    this._clientFactory = clientFactory;
+    this._key = key;
+  }
+
+  /**
+   * Sends the event to the passed in telemetry client
+   * @param eventName The name of the event to log
+   * @param data The data contained in the event
+   */
+  sendEventData(eventName: string, data?: SenderData): void {
+    if (!this._telemetryClient) {
+      if (this._instantiationStatus !== InstantiationStatus.INSTANTIATED) {
+        this._eventQueue.push({ eventName, data });
+      }
+      return;
+    }
+    this._telemetryClient.logEvent(eventName, data);
+  }
+
+  /**
+   * Sends an exception to the passed in telemetry client
+   * @param exception The exception to collect
+   * @param data Data associated with the exception
+   */
+  sendExceptionData(exception: Error, data?: SenderData): void {
+    if (!this._telemetryClient) {
+      if (this._instantiationStatus !== InstantiationStatus.INSTANTIATED) {
+        this._exceptionQueue.push({ exception, data });
+      }
+      return;
+    }
+    this._telemetryClient.logException(exception, data);
+  }
+
+  /**
+   * Flushes the buffered telemetry data
+   */
+  async flush(): Promise<void> {
+    if (this._telemetryClient) {
+      await this._telemetryClient.flush();
+      this._telemetryClient = undefined;
```

Comment:
are you sure that we need to set the telemetryclient to undefined after flush? can we call it only once?